Sources/LeafKit/LeafLexer.swift

```diff
@@ -132,123 +141,191 @@ struct LeafLexer {
-            case let x where x == .exclamation
-                && src.peek(aheadBy: 1)?.isValidOperator == false:
-                // for when ! stands alone
-                src.pop()
-                return .parameter(.operator(.not))
-            case let x where x.isValidInParameter:
-                let read = src.readWhile { $0.isValidInParameter }
-                guard let name = read else { fatalError("disallowed by switch") }
-                // this parameter is a tag
-                if src.peek() == .leftParenthesis { return .parameter(.tag(name: name)) }
-                
-                // check if expected parameter type
-                if let keyword = Keyword(rawValue: name) { return .parameter(.keyword(keyword)) }
-                else if let op = Operator(rawValue: name) { return .parameter(.operator(op)) }
-                else if let val = Int(name) { return .parameter(.constant(.int(val))) }
-                else if let val = Double(name) { return .parameter(.constant(.double(val))) }
-                
-                // unknown param type.. var
-                return .parameter(.variable(name: name))
-            default:
-                throw LexerError(src: src, lexed: lexed, reason: .invalidParameterToken(next))
+                let read = src.readWhile { $0 == .space } ?? ""
+                return .whitespace(length: read.count + 1)
+            default: break
+        }
+        
+        // Complex Parameter lexing situations - enhanced to allow non-whitespace separated values
+        // Complicated by overlap in acceptable isValidInParameter characters between possible types
+        // Process from most restrictive options to least to help prevent overly aggressive tokens
+        // Possible results, most restrictive to least
+        // * Operator
+        // * Constant(Int)
+        // * Constant(Double)
+        // * Keyword
+        // * Tag
+        // * Variable
+        
+        // if current character isn't valid for any kind of parameter, something's majorly wrong
+        guard current.isValidInParameter else {
+            throw LexerError(.invalidParameterToken(current), src: src, lexed: lexed)
+        }
+
+        // Test for Operator first - this will only handle max two character operators, not ideal
+        // Can't switch on this, MUST happen before trying to read tags
+        if current.isValidOperator {
+            var op = Operator(rawValue: String(current) + String(src.peek()!))
+            if op == nil { op = Operator(rawValue: String(current)) } else { src.pop() }
+            if op != nil { return .parameter(.operator(op!)) }
+        }
+        
+        // Test for numerics next. This is not very intelligent but will read base2/8/10/16
+        // for Ints and base 10/16 for decimal through native Swift initialization
+        // Will not adequately decay to handle things like `0b0A` and recognize as invalid.
+        if current.canStartNumeric {
```

Comment:
### BREAKING REGRESSION ###
While the new lexing logic here improves a lot of things about handling constants and operators, this introduces a small problem that can't be handled without a look-behind structure. This is easily solved in Parser by properly compressing operator(minus) as a prefix unary onto the constant, and is better solved there than in Lexer, IMO
Input | Expectation | Correct Before? | Correct Now?
-- | -- | -- | --
`index-5` | `variable(index) operator(-) constant(5)` | No - `variable(index-5)` | Yes
`10-5` | `constant(10) operator(-) constant(5)` | No - `variable(10-5)` | Yes
`10 - 5` | `constant(10) operator(-) constant(5)` | Yes | Yes
`10 - -5` | `constant(10) operator(-) constant(-5)` | Yes | No - `constant(10) operator(-) operator(-) constant(5)`
`-5` | `constant(-5)` | No - `raw(-5)` | No - `operator(-) constant(5)`
`- 5` | `operator(-) constant(5) | Yes | No - `operator(-) constant(5)`

Reply:
New commit resolves negatives. Note that this comment was *slightly* inaccurate as the last case should not be resolvable as a prefix when whitespace is present, and I'm not sure why I said it wasn't correct in this PR since operator + constant are the correct behaviors.